stdlib/mexpr/parser.mc

```diff
@@ -0,0 +1,433 @@
+end
+
+-- Eats whitespace
+lang WhitespaceParser = WSACParser
+  sem eatWSAC (p : Pos)  =
+  | " " ++ xs -> eatWSAC (advanceCol p 1)  xs
+  | "\t" ++ xs -> eatWSAC (advanceCol p tabSpace) xs
+  | "\n" ++ xs -> eatWSAC (advanceRow p 1) xs
+  | "\r" ++ xs -> eatWSAC p xs
+  | x -> {str = x, pos = p}
+end
+
+let _ = use WhitespaceParser in
+  utest eatWSAC (initPos "") "  foo" with {str = "foo", pos = (posVal "" 1 2)} in
+  utest eatWSAC (initPos "") " \tfoo" with {str = "foo", pos = (posVal "" 1 3)} in
+  utest eatWSAC (initPos "") " \n    bar " with {str = "bar ", pos = (posVal "" 2 4)} in
+  ()
+
+-- Eat line comments of the form --
+lang LineCommentParser = WSACParser
+  sem eatWSAC (p : Pos)  =
+  | "--" ++ xs ->
+    recursive
+    let remove = lam p. lam str.
+      match str with "\n" ++ xs then eatWSAC (advanceRow p 1) xs else
+      match str with [x] ++ xs then remove (advanceCol p 1) xs else
+      eatWSAC p str
+    in remove p xs
+end
+
+-- Eat multiline comment of the form *-  -*
+lang MultilineCommentParser = WSACParser
+  sem eatWSAC (p : Pos) =
+  | "/-" ++ xs ->
+    recursive
+    let remove = lam p. lam str. lam d.
+      match str with "/-" ++ xs then remove (advanceCol p 2) xs (addi d 1) else
+      match str with "\n" ++ xs then remove (advanceRow p 1) xs d else
+      match str with "-/" ++ xs then
+        if eqi d 1 then eatWSAC (advanceCol p 2) xs
+        else remove (advanceCol p 2) xs (subi d 1) else
+      match str with [_] ++ xs then remove (advanceCol p 1) xs d else
+      if eqi d 0 then eatWSAC p str else posErrorExit p "Unmatched multiline comments."
+    in remove (advanceCol p 2) xs 1
+end
+
+-- Commbined WSAC parser for MExpr
+lang MExprWSACParser = WhitespaceParser + LineCommentParser + MultilineCommentParser
+
+let _ = use MExprWSACParser in
+  utest eatWSAC (initPos "") " --foo \n  bar " with {str = "bar ", pos = posVal "" 2 2} in
+  utest eatWSAC (initPos "") " /- foo -/ bar" with {str = "bar", pos = posVal "" 1 11} in
+  utest eatWSAC (initPos "") " /- foo\n x \n -/ \nbar " with {str = "bar ", pos = posVal "" 4 0} in
+  utest eatWSAC (initPos "") " /- x -- y /- foo \n -/ -/ !" with {str = "!", pos = posVal "" 2 7} in
+  ()
+
+
+
+-- Top of the expression parser. Connects WSAC with parsing of other non terminals
+lang ExprParser = WSACParser
+  sem parseExpr (p: Pos) =
+  | s ->
+    let r1 = parseExprMain p 0 s in
+    let r2 = eatWSAC r1.pos r1.str in
+    if eqi (length r2.str) 0 then r1.val
+    else posErrorExit r2.pos "Parse error. Unknown characters."
+  
+  sem parseExprMain (p: Pos) (prec: Int) =
+  | s ->
+    let r1 = eatWSAC p s in
+    let exp = parseExprImp r1.pos r1.str in
+    let r2 = eatWSAC exp.pos exp.str in
+    parseInfix r2.pos prec exp r2.str 
+
+  sem parseInfix (p: Pos) (prec: Int) (exp: Expr) =
+
+  sem parseExprImp (p: Pos) =
+  | _ -> posErrorExit p "Parse error. Unknown character sequence."
+end
+
+
+-- Include this fragment is there are no infix operations
```

Comment:
Is this "include if" or "include iff"? If it's the latter it's a bit awkward that we then can't add infix operators to a language that lacks them, since it will already have this fragment included.

Reply:
Yes, this how it has to be done with the current design in MLang. If we extend MLang patterns to make a difference between pattern

```
| e -> ...
```

and 
```
| _ -> ...
```
where the former has higher priority, we can make a better design. I would suggest that we add this to MLang patterns.